ml-agents/mlagents/trainers/saver/saver.py

```diff
@@ -0,0 +1,39 @@
+# # Unity ML-Agents Toolkit
+import abc
+from mlagents.trainers.policy import Policy
+
+
+class BaseSaver(abc.ABC):
+    """This class is the base class for the Saver"""
+
+    def __init__(self):
+        pass
+
+    @abc.abstractmethod
+    def register(self, module: Any, init_or_load: bool) -> None:
```

Comment:
I still think you should be explicit about the type of thing you're registering; checking the type of an object and branching on that is generally an anti-pattern. 

You can make them non-abstract and do nothing by default, in case one of the implementations doesn't need to do anything (like it appears TFSave ignores optimizers)

Reply:
What if we have separate `register_policy` and `register_optimizer` methods? That should make the typing explicit.

The branching on `TFPolicy` is just for TensorFlow since the policy object holds the graph and sess. In PyTorch we can treat the registered classes in the same way. In that world we could make an interface type (`SerializableModel` or similar) and set the typing to that here.

Reply:
The awkwardness here might come from the fact that this is designed to work for both TF and Torch which are doing checkpointing in very different way. In Torch, everything that has a `.get_modules()` function can be registered here and optimizer is not skipped. 

Will make it non-abstract.

Reply:
So do we want this function to have fixed input of only accepting policy and optimizer?
This is totally fine for tensorflow, and it's ok for current pytorch.
If we're not going to add other type of objects in pytorch I can change this to `register_policy` and `register_optimizer`.

Reply:
Add specified type as Union[Policy, Optimizer] for this. This should clear the concern of unclear type.

Reply:
To make it a bit cleaner, what if we do (in the TF implementation):
```
def register(self, module: Union[Policy, Optimizer]):
    if isinstance(module, Optimizer):
        self._register_optimizer(module)
    elif isinstance(module, Policy):
        self._register_optimizer(module)
    else: 
        # throw some error
```
The two helper methods can have Optimizer and Policy input types, and are probably not needed in the Torch implementation.

Reply:
Yes as this is probably not needed in the Torch, I'll put it in TFSaver.